src/main/java/core/basesyntax/ArrayList.java

```
@@ -1,48 +1,129 @@
+            for (int i = 0; i < list.size(); i++) {
+                add(list.get(i));
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        T result = getting(index);
+        if (result != null && result.getClass().getGenericSuperclass() == null) {
+            throw new ArrayListIndexOutOfBoundsException(OF_BOUND_MESSAGE);
+        }
+        return result;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!setting(value, index)) {
+            throw new ArrayListIndexOutOfBoundsException(OF_BOUND_MESSAGE);
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T result = removing(index);
+        if (result == null) {
+            throw new ArrayListIndexOutOfBoundsException(OF_BOUND_MESSAGE);
+        }
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(storage[i], element)) {
+                return removing(i);
+            }
+        }
+        throw new NoSuchElementException(OF_ABSENT_MESSAGE);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean adding(T value, int index) {
+        if (index >= 0 && index <= size) {
+            List<T> tempList = null;
+            if (index < size) {
+                tempList = new ArrayList<>();
+                for (int i = index; i < size; i++) {
+                    tempList.add(get(i));
+                }
+            }
+            size = index;
+            add(value);
+            addAll(tempList);
+        }
+        return index >= 0 && index < size;
+    }
+
+    private T getting(int index) {
+        return index < 0 || index >= size ? (T) new Object() : storage[index];
+    }
```

Why are you returning `new Object` in case of an invalid index? we should throw an exception right away? Use a `checkIndex()` method which I have described above

---

refactored